### PR TITLE
feat(memory): Integrate NibblePack-compression into serialized Histograms

### PR DESCRIPTION
**Pull request by velvia**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

Histograms sent over the wire are uncompressed.

**New behavior :**

Integrate the NibblePacking compression algorithm into the `BinaryHistogram` format used for encoding histograms and buckets over the wire (Kafka and query results).  Saves significant space.

HistogramColumn vector format is still columnar however and unsuitable for persistence.

### CODE REVIEW

**Comment by velvia** [comment]
@broneill @vishramachandran please review this one next, NibblePacking changes have been merged in here.  It is simply integration of NibblePacking into histogram code.

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala [hunk 1]
```diff
@@ -119,24 +120,49 @@ object Histogram {
...
   }
 }
 
+trait HistogramWithBuckets extends Histogram {
+  def buckets: HistogramBuckets
+  final def numBuckets: Int = buckets.numBuckets
+  final def bucketTop(no: Int): Double = buckets.bucketTop(no)
+}
+
+final case class LongHistogram(buckets: HistogramBuckets, values: Array[Long]) extends HistogramWithBuckets {
```
Comment (vishramachandran) [inline memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala]: This PR too does not have the annotation on whether the histogram is based on counter based bucket values or increase based bucket values. This is important - I had called this out in earlier PR.

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala [hunk 2]
```diff
@@ -119,24 +120,49 @@ object Histogram {
...
+  final def bucketValue(no: Int): Double = values(no).toDouble
+  final def serialize(intoBuf: Option[MutableDirectBuffer] = None): MutableDirectBuffer = {
+    val buf = intoBuf.getOrElse(BinaryHistogram.histBuf)
+    buckets match {
+      case g: GeometricBuckets => BinaryHistogram.writeDelta(g, values, buf)
+    }
```
[the remark this answers is not in the document]
Comment (velvia) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala]: For now, yes.  When I add support for user-specified buckets then they won't be the only ones.

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala (hunk 1 above)
Comment (velvia) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala]: Yeah, I still remember that.  It doesn't belong in each histogram though.  The counter vs increase is a property of the entire column, not each histogram.  I'll think about where to add that and the scope.

**Pull request merged**